client/cmd/job/plan.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/goto/optimus/client/extension/provider/github"
+	"github.com/goto/optimus/client/extension/provider/gitlab"
+	"github.com/goto/optimus/client/local"
+	"github.com/goto/optimus/client/local/model"
+	"github.com/goto/optimus/client/local/specio"
+	"github.com/goto/optimus/config"
+)
+
+const jobFileName = "job.yaml"
+
+type planCommand struct {
+	logger log.Logger
+
+	sourceRef, destinationRef     string
+	gitURL, gitToken, gitProvider string
+	verbose                       bool
+	projectID                     string
+	output                        string
+
+	configFilePath string
+	clientConfig   *config.ClientConfig
+
+	repository     providermodel.RepositoryAPI
+	specReadWriter local.SpecReadWriter[*model.JobSpec]
+}
+
+func NewPlanCommand() *cobra.Command {
+	planCmd := &planCommand{logger: logger.NewClientLogger()}
+	cmd := &cobra.Command{
+		Use:     "plan",
+		Short:   "Plan job deployment",
+		Long:    "Plan job deployment based on git diff state using git reference (commit SHA, branch, tag)",
+		Example: "optimus job plan <ref> <ref-before>",
+		Args:    cobra.MinimumNArgs(2), //nolint
+		PreRunE: planCmd.PreRunE,
+		RunE:    planCmd.RunE,
+	}
+
+	planCmd.inject(cmd)
+	return cmd
+}
+
+func (p *planCommand) inject(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&p.configFilePath, "config", "c", config.EmptyPath, "File path for client configuration")
+
+	cmd.Flags().StringVarP(&p.gitProvider, "git-provider", "p", os.Getenv("GIT_PROVIDER"), "selected git provider used in the repository")
+	cmd.Flags().StringVarP(&p.gitURL, "git-host", "h", os.Getenv("GIT_HOST"), "Git host based on git provider used in the repository")
+	cmd.Flags().StringVarP(&p.gitToken, "git-token", "t", os.Getenv("GIT_TOKEN"), "Git token based on git provider used in the repository")
+
+	cmd.Flags().StringVarP(&p.projectID, "project-id", "I", os.Getenv("GIT_PROJECT_ID"), "Determine which project will be checked")
+
+	cmd.Flags().StringVarP(&p.output, "output", "o", "./job.json", "File path for output of plan")
+	cmd.Flags().BoolVarP(&p.verbose, "verbose", "v", false, "Print details related to operation")
+}
+
+func (p *planCommand) PreRunE(_ *cobra.Command, args []string) error {
+	var err error
+
+	p.destinationRef = args[0]
+	p.sourceRef = args[1]
+	p.logger.Info("[plan] compare: `%s` ← `%s`", p.destinationRef, p.sourceRef)
+	p.clientConfig, err = config.LoadClientConfig(p.configFilePath)
+	if err != nil {
+		return err
+	}
+
+	switch p.gitProvider {
+	case "gitlab":
+		p.repository, err = gitlab.NewAPI(p.gitURL, p.gitToken)
+	case "github":
+		p.repository, err = github.NewAPI(p.gitURL, p.gitToken)
+	default:
+		return errors.New("unsupported git provider, we currently only support: [github,gitlab]")
+	}
+	if err != nil {
+		return err
+	}
+
+	p.specReadWriter, err = specio.NewJobSpecReadWriter(afero.NewOsFs())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *planCommand) RunE(_ *cobra.Command, _ []string) error {
```

Comment:
As discussed imo, plan command should not have the logic of plan, it should just invoke the functionality implemented in the plan struct.

Reply:
and see if we can create smaller functions & tie them together